metacal.py

```diff
@@ -0,0 +1,288 @@
+    g1_1p = data['g'][w_1p, 0].mean()
+    g1_1m = data['g'][w_1m, 0].mean()
+    R11 = (g1_1p - g1_1m)/0.02
+
+    shear = g / R11
+    shear_err = gerr / R11
+
+    m = shear[0]/shear_true[0]-1
+    merr = shear_err[0]/shear_true[0]
+
+    s2n = data['s2n'][w].mean()
+
+    print('sigma_n: %g' % sigma_n)
+    print('Metacalibration')
+    print('----------------------')
+    print('S/N: %g' % s2n)
+    print('R11: %g' % R11)
+    print('shear', shear)
+    print('shear true', shear_true)
+    print('m: %g +/- %g (99.7%% conf)' % (m, merr*3))
+    print('c: %g +/- %g (99.7%% conf)' % (shear[1], shear_err[1]*3))
+
+
+def select(data, shear_type):
+    """
+    select the data by shear type and size
+
+    Parameters
+    ----------
+    data: array
+        The array with fields shear_type and T
+    shear_type: str
+        e.g. 'noshear', '1p', etc.
+
+    Returns
+    -------
+    array of indices
+    """
+
+    w, = np.where(
+        (data['flags'] == 0) & (data['shear_type'] == shear_type)
+    )
+    return w
+
+
+def make_struct(res, obs, shear_type):
+    """
+    make the data structure
+
+    Parameters
+    ----------
+    res: dict
+        With keys 's2n', 'e', and 'T'
+    obs: ngmix.Observation
+        The observation for this shear type
+    shear_type: str
+        The shear type
+
+    Returns
+    -------
+    1-element array with fields
+    """
+    dt = [
+        ('flags', 'i4'),
+        ('shear_type', 'U7'),
+        ('s2n', 'f8'),
+        ('g', 'f8', 2),
+        ('T', 'f8'),
+        ('Tpsf', 'f8'),
+    ]
+    data = np.zeros(1, dtype=dt)
+    data['shear_type'] = shear_type
+    data['flags'] = res['flags']
+    if res['flags'] == 0:
+        data['s2n'] = res['s2n']
+        # for moments we are actually measureing e, the elliptity
+        data['g'] = res['e']
+        data['T'] = res['T']
+    else:
+        data['s2n'] = np.nan
+        data['g'] = np.nan
+        data['T'] = np.nan
+        data['Tpsf'] = np.nan
+
+        # we only have one epoch and band, so we can get the psf T from the
+        # observation rather than averaging over epochs/bands
+        data['Tpsf'] = obs.psf.meta['result']['T']
+
+    return data
+
+cat = galsim.COSMOSCatalog()
+
+#def make_data(stamp, rng, noise, shear):
+def make_data(data, i, sigma_n):
+    """
+    simulate an exponential object with moffat psf
+
+    Parameters
+    ----------
+    rng: np.random.RandomState
+        The random number generator
+    noise: float
+        Noise for the image
+    shear: (g1, g2)
+        The shear in each component
+
+    Returns
+    -------
+    ngmix.Observation
+    """
+    scale = _scale
+
+    psf = cat.makeGalaxy(2,  gal_type='real', noise_pad_size=0).original_psf
+    
+    # import here simulations
+    im = data[i,...]
+
+    noise = sigma_n
+    wt = im*0 + 1.0/noise**2
+
+    psf_im = psf.drawImage(scale=scale).array
+
+    psf_obs = ngmix.Observation(
+        psf_im,
+    )
+
+    obs = ngmix.Observation(
```

Comment:
@b-remy When you create obs you should specify the pixel scale, I think this might be one of the sources of the issues.

see here: https://github.com/CosmoStat/autometacal/blob/f5958c41c091b229017be5f072f1c58b94120d80/examples/metacal.py#L296

You can probably remove the mis-centering question.

Reply:
Exact. I added psf and image jacobians in my code. But I don't get to which transformation these jacobians correspond. Do you have an idea?

Reply:
When you only specify the pixel scale, the Jacobian is purely diagonal. No transform is applied to the image, it simply indicates how you go from pixel units to arcsecs